### PR TITLE
PyTorch Synthetic Benchmark

### PR DESCRIPTION
**Pull request by alsrgv**

### CODE REVIEW

File: examples/pytorch_synthetic_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+# Horovod: wrap optimizer with DistributedOptimizer.
+optimizer = hvd.DistributedOptimizer(optimizer,
+                                     named_parameters=model.named_parameters(),
+                                     compression=compression)
+
+# Horovod: broadcast parameters & optimizer state.
+hvd.broadcast_parameters(model.state_dict(), root_rank=0)
+# TODO: needs bugfix
+#hvd.broadcast_optimizer_state(optimizer, root_rank=0)
```
Comment (alsrgv) [inline examples/pytorch_synthetic_benchmark.py]: @tgaddair, we should fix this bug before landing.  `optim.SGD` w/o momentum & weight decay cause the following issue with `broadcast_optimizer_state`:
```
Traceback (most recent call last):
  File "pytorch_synthetic_benchmark.py", line 64, in <module>
    hvd.broadcast_optimizer_state(optimizer, root_rank=0)
  File "/usr/local/lib/python2.7/site-packages/horovod/torch/__init__.py", line 213, in broadcast_optimizer_state
    param_state = state_dict['state'][pid]
KeyError: 4592867280
```

Comment (tgaddair) [reply examples/pytorch_synthetic_benchmark.py]: Done #548.

**Comment by bapriddy** [comment]
@alsrgv I'm getting the following error when running the pytorch_synthetic_benchmark.py

` Traceback (most recent call last):
  File "syn.py", line 64, in <module>
    hvd.broadcast_optimizer_state(optimizer, root_rank=0)
  File "~/miniconda3/lib/python3.6/site-packages/horovod/torch/__init__.py", line 213, in broadcast_optimizer_state
    param_state = state_dict['state'][pid]
KeyError: 140230636548384 `

I get the error in parallel as well.  Same output k times...For example 

` aprun -n 4 -N 1 ~/miniconda3/bin/python syn.py `

Gives 
` Traceback (most recent call last):
  File "syn.py", line 64, in <module>
Traceback (most recent call last):
  File "syn.py", line 64, in <module>
Traceback (most recent call last):
  File "syn.py", line 64, in <module>
Traceback (most recent call last):
  File "syn.py", line 64, in <module>
    hvd.broadcast_optimizer_state(optimizer, root_rank=0)
  File "~/miniconda3/lib/python3.6/site-packages/horovod/torch/__init__.py", line 213, in broadcast_optimizer_state
    hvd.broadcast_optimizer_state(optimizer, root_rank=0)
  File "~/miniconda3/lib/python3.6/site-packages/horovod/torch/__init__.py", line 213, in broadcast_optimizer_state
    hvd.broadcast_optimizer_state(optimizer, root_rank=0)
  File "~/miniconda3/lib/python3.6/site-packages/horovod/torch/__init__.py", line 213, in broadcast_optimizer_state
    hvd.broadcast_optimizer_state(optimizer, root_rank=0)
  File "~/miniconda3/lib/python3.6/site-packages/horovod/torch/__init__.py", line 213, in broadcast_optimizer_state
    param_state = state_dict['state'][pid]
    param_state = state_dict['state'][pid]
KeyError: 46913509835904
KeyError: 46913509835904
    param_state = state_dict['state'][pid]
KeyError: 46913509835904
    param_state = state_dict['state'][pid]
KeyError: 46913509835904 `


I've included the reference call below in /torch/__init__.py

` def _create_callback(pid, name, t, p): `
        `def _from_tensor():`
           ` state_dict['state'][pid][name] = t(p.numpy()[0])`
      `  return _from_tensor `

**Comment by bapriddy** [comment]
Yes.  I'll check it.  Thanks.

**Comment by bapriddy** [comment]
It worked.  Thanks!!

**Comment by bapriddy** [comment]
@alsrgv How does the code decide when to stop "warmup" and proceed with the test??  Just curious.  Also, would switching to fp16 have any effect?

**Comment by alsrgv** [comment]
@bapriddy, warmup runs for `--num-warmup-batches`, which defaults to 10.  Using `--fp16-allreduce` should improve performance if your network is slow.

**Comment by alsrgv** [comment]
@bapriddy, warmup runs for `--num-warmup-batches`, which defaults to 10.  Using `--fp16-allreduce` should improve performance if your network is slow.

**Comment by alsrgv** [comment]
@bapriddy, warmup runs for `--num-warmup-batches`, which defaults to 10.  Using `--fp16-allreduce` should improve performance if your network is slow.

**Comment by alsrgv** [comment]
@bapriddy, warmup runs for `--num-warmup-batches`, which defaults to 10.  Using `--fp16-allreduce` should improve performance if your network is slow.

**Comment by alsrgv** [comment]
@bapriddy, warmup runs for `--num-warmup-batches`, which defaults to 10.  Using `--fp16-allreduce` should improve performance if your network is slow.

**Comment by alsrgv** [comment]
@bapriddy, warmup runs for `--num-warmup-batches`, which defaults to 10.  Using `--fp16-allreduce` should improve performance if your network is slow.

**Comment by alsrgv** [comment]
@bapriddy, warmup runs for `--num-warmup-batches`, which defaults to 10.  Using `--fp16-allreduce` should improve performance if your network is slow.

**Comment by alsrgv** [comment]
@bapriddy, warmup runs for `--num-warmup-batches`, which defaults to 10.  Using `--fp16-allreduce` should improve performance if your network is slow.

**Comment by alsrgv** [comment]
@bapriddy, warmup runs for `--num-warmup-batches`, which defaults to 10.  Using `--fp16-allreduce` should improve performance if your network is slow.

**Comment by alsrgv** [comment]
@bapriddy, warmup runs for `--num-warmup-batches`, which defaults to 10.  Using `--fp16-allreduce` should improve performance if your network is slow.

**Comment by alsrgv** [comment]
@bapriddy, warmup runs for `--num-warmup-batches`, which defaults to 10.  Using `--fp16-allreduce` should improve performance if your network is slow.

**Comment by bapriddy** [comment]
@alsrgv  Thanks.  How can I modify pytorch_synthetic_benchmark.py to run resnet18, resnet110, etc?

**Comment by bapriddy** [comment]
@alsrgv   Is it possible to modify the pytorch_synthetic_benchmark.py for resnet18, resnet101, or other imagenet models?   I did this with pytorch_imagenet_resnet50.py by changing line 114.

` # Set up standard ResNet-50 model. ` 
`  model = models.resnet50() `

**Comment by alsrgv** [comment]
@bapriddy, yeah, you can just pass `--model resnet101`.

**Comment by bapriddy** [comment]
@alsrgv Got it!

**Pull request merged**